docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
 
 Another important difference between type variables such as ``AnyStr``
 and using ``@overload`` is that the prior can also be used to define
-constraints for generic class type parameters.  For example, the type
+constraints for generic class type parameters. For example, the type
 parameter of the generic class ``typing.IO`` is constrained (only
 ``IO[str]``, ``IO[bytes]`` and ``IO[Any]`` are valid)::
 
   class IO(Generic[AnyStr]): ...
+
+
+Invalid overload definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should enforce the following rules for overload definitions.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+An overload implementation should be present for all overload function
+definitions. Type checkers should report an error or warning if an
+implementation is missing. Overload definitions within stub files,
+protocols, and abstract base classes are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Similarly, if one overload
+is ``async``, all overloads must be ``async`` as well as the implementation.
+Type checkers should report an error if these conditions are not met.
```

Comment:
This could stand to be slightly clearer in that it refers to coroutine functions defined using async def, and not all async functions.

A concrete example of an API that is allowed which could be considered async with one overload defintion, but not a coroutine function for contrast

```py
@overload
def get(self, *, wait: Literal[True]) -> Awaitable[T]:  ...
@overload
def get(self, *, wait: Literal[False]) -> T:  ...
def get(self, * wait: bool = True) -> Awaitable[T] | T: ...
```

Reply:
We should take care to be clear that `def` overloads + `async def` implementation is fine (this makes some cases with async generators work better, see also https://discuss.python.org/t/overloads-of-async-generators-inconsistent-coroutine-wrapping/56665/9).

It might also be useful for the spec to explicitly mention that `yield` in the implementation body does not affect types of the overloads.

Reply:
Based on @hauntsaninja's point, it sounds like there is a desire not to enforce `async def` consistency for overloads. I'll delete that point and make it clear that a mixture of `async def` and `def` is supported.